### PR TITLE
Optimize load time

### PR DESCRIPTION
**Pull request by Glease**

### CODE REVIEW

File: src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java [hunk 1]
```diff
@@ -28,11 +28,11 @@
 import com.github.bartimaeusnek.bartworks.system.material.WerkstoffLoader;
 import com.github.bartimaeusnek.bartworks.system.material.processingLoaders.AfterLuVTierEnhacement;
 import com.github.bartimaeusnek.bartworks.util.BW_Util;
-import gregtech.api.enums.ItemList;
-import gregtech.api.enums.Materials;
-import gregtech.api.enums.OrePrefixes;
+import gregtech.api.enums.*;
```
Comment (bombcar) [inline src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java:31]: Coding style question: are we for or against the .*

After all most of the apis are five years stable now ….

Comment (Glease) [reply src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java:31]: If no repo owner (e.g. me to TC4Tweaks, mitch to NEI, etc) objects then use whatever IDE adds for you.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java [hunk 2]
```diff
@@ -135,22 +136,26 @@ public static void unificationRecipeEnforcer() {
...
+                                    GT_Recipe recipe = iterator.next();
+                                    boolean removal = map.equals(GT_Recipe_Map.sFluidExtractionRecipes) || map.equals(GT_Recipe_Map.sFluidSolidficationRecipes);
                                     for (int i = 0; i < recipe.mInputs.length; i++) {
                                         if (!GT_Utility.areStacksEqual(recipe.mInputs[i], toReplace))
                                             continue;
-                                        if (removal)
-                                            toRem.add(recipe);
+                                        if (removal) {
+                                            iterator.remove();
+                                            continue nextRecipe;
```
Comment (TheElan) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java:157]: It's probably better to extract inner loop to a method then use labels

File: src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java [hunk 3]
```diff
@@ -93,20 +93,25 @@ public void run() {
...
+                    rewriteMachineRecipes(stack, LuVMaterialsGenerated, recipe);
                 }
-        );
+            }
+        }
+
+        for (ItemStack stack : LuVMachines) {
+            Predicate recipeFilter = obj -> obj instanceof GT_Shaped_Recipe && GT_Utility.areStacksEqual(((GT_Shaped_Recipe) obj).getRecipeOutput(), stack, true);
+            rewriteCraftingRecipes(bufferedRecipeList, LuVMaterialsGenerated, recipeFilter);
+        }
```
Comment (TheElan) [inline src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java:114]: Couldn't this be reduced to one for each on each LuV machine and doing the logic in one pass?

Btw how much overhead streams have caused here?

**Review by TheElan — COMMENTED** [review]
Having improvement metrics on each change or at least on overall change would be nice

File: src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java (hunk 2 above)
Comment (mitchej123) [reply src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java:157]: I think this is clear, rewriting without labels and adding more functional overhead wouldn't help much imho.

File: src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java (hunk 1 above)
Comment (mitchej123) [reply src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java:31]: I set my IDE to disambiguate the imports, and I prefer explicit... but won't get too pedantic about it outside of like NEI as mentioned.

**Review by mitchej123 — APPROVED** [review]
Would be nice to see some profiling before and after, but other than that the changes look reasonable.  I noticed some of these same areas showing up on my profiles to.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java (hunk 2 above)
Comment (TheElan) [reply src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java:157]: Labels give me a bit of a `goto` vibes so I prefer alternatives. Besides moving things into a method shouldn't result in drastic overhead if any (compiler can inline things where it sees fit)

File: src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java (hunk 3 above)
Comment (Glease) [reply src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java:114]: > Couldn't this be reduced to one for each on each LuV machine and doing the logic in one pass?

That's exactly what bart was doing. It certainly looks cleaner but has terrible performance. This change cuts 6 second load time into 1 second.

> Btw how much overhead streams have caused here?

Not quite sure. Didn't had time to measure it.

**Comment by Glease** [comment]
I do have the profiler outputs. I forgot to add them to the PR though.

## Changes in cached reflection:
Before:
before

After:
after



## Changes in LuVTierEnhancer

Before:
Before

After:
After

Extended reading: [Cache Blocking](https://www.intel.com/content/www/us/en/developer/articles/technical/cache-blocking-techniques.html)

**Comment by mitchej123** [comment]
Nice!

File: src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java [hunk 4]
```diff
@@ -135,22 +136,26 @@ public static void unificationRecipeEnforcer() {
                     for (OrePrefixes prefixes : OrePrefixes.values()) {
                         if (!werkstoff.hasItemType(prefixes))
                             continue;
-                        if (OreDictionary.getOres(prefixes + s).size() <= 1)
+                        String fullOreName = prefixes + s;
+                        List<ItemStack> ores = OreDictionary.getOres(fullOreName, false);
+                        if (ores.isEmpty())
```
Comment (Skiftcha) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java:141]: || size == 1?

File: src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java (hunk 3 above)
Comment (TheElan) [reply src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java:114]: Interesting, how did you come to conclusion that you want to switch nesting?

Comment (TheElan) [reply src/main/java/com/github/bartimaeusnek/bartworks/system/material/GT_Enhancement/LuVTierEnhancer.java:114]: Ah, ok you've provided link below, I get it now

File: src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java (hunk 4 above)
Comment (Glease) [reply src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java:141]: oops. Good catch.

Comment (Glease) [reply src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/StaticRecipeChangeLoaders.java:141]: Functionally it doesn't make it behave in wrong ways (1 ore => no need to unification, so continue, but not continuing won't hurt either) I'll patch this quickly...

**Comment by bartimaeusnek** [comment]
@Glease good stuff 👍

**Pull request merged**